new tool for AmberTools 'tleap'

thepineapplepirate opened the pull request:
This is a new tool for the AmberTools program "tleap". This tool will fulfill many needs, including now having full access to the newer AMBER force fields, and other specialized forcefields (metalloprotein forcefields like ZAFF and EZAFF, for example), which are not in the GROMACS source code. By generating these files with tLeap, and then processing the outputs with acpype, these newer features can successfully be used with MD simulations conducted in Galaxy/GROMACS.

bgruening commented on tools/buildtools/ambertools/tleap.xml:
```diff
@@ -0,0 +1,2043 @@
+    <description>interactively build and run tLEaP files to set up systems with AmberTools</description>
+     <macros>
+        <import>macros.xml</import>
+        <token name="@GALAXY_VERSION@">0</token>
```
token is not used?

Should the version be taken from the macro?

thepineapplepirate replied:
corrected

bgruening commented on tools/buildtools/ambertools/tleap.xml:
```diff
@@ -0,0 +1,2043 @@
+        <token name="@GALAXY_VERSION@">0</token>
+     </macros>
+     <expand macro="requirements" />
+     <stdio>
```
remove this stdio and replace the functionality with detect_errors https://docs.galaxyproject.org/en/latest/dev/schema.html#id24

bgruening commented on tools/buildtools/ambertools/tleap.xml:
```diff
@@ -0,0 +1,2043 @@
+            <param name="arg_solvateDontClip_arg_5" label="solvateDontClip_arg_5 value" type="text" value="" help=""/>
+       </when>
+       <when value="solvateOct">
+            <param name="arg_solute" label="solute value" type="text" value="" help="UNIT"/>
```
Would it make sense to use way more macros here? Those params are used a lot everywhere.

thepineapplepirate replied:
corrected

bgruening commented on tools/buildtools/ambertools/tleap.xml:
```diff
@@ -0,0 +1,2043 @@
+                    <option value="builtin">From Built-in AmberTools data</option>
+               </param>
+               <when value="history">
+                    <param name="arg_filename" label="File name" value="" type="data" format="txt"/>
```
its not the name, its the file

thepineapplepirate replied:
corrected

bgruening commented on tools/buildtools/ambertools/tleap.xml:
```diff
@@ -0,0 +1,2043 @@
+       </when>
+       <when value="loadAmberParams">
+            <param name="loadAmberParams_assign" label="Assign to variable named" type="text" value=""/>
+            <conditional name="file_source">
```
entire conditional into the macro?

thepineapplepirate replied:
corrected

bgruening commented on tools/buildtools/ambertools/tleap.xml:
```diff
@@ -0,0 +1,2043 @@
+            <param name="arg_obj" label="obj value" type="text" value="" help="UNIT"/>
+       </when>
+       <when value="groupSelectedAtoms">
+            <param name="arg_unit" label="unit value" type="text" value="" help="UNIT"/>
```
macros?

thepineapplepirate replied:
> Cool, thanks!
> 
> ![grafik](https://user-images.githubusercontent.com/469983/149353685-6ed1e239-6cad-4ab1-8d65-7f43b0aa2041.png)

nice man!!! right on

thepineapplepirate replied:
corrected

bgruening commented on tools/buildtools/ambertools/tleap.xml:
```diff
@@ -0,0 +1,2043 @@
+          <repeat name="tleap_cmds">
+            <conditional name="tleap_cond">
+                 <param name="tleap_cmd" value="source"/>
+                   <param name="arg_filename" value="oldff/leaprc.ff14SB"/>
```
indentation is off here ... do you know about https://github.com/galaxyproject/galaxy-language-server ?

It will help you with style but also with macros etc ...

thepineapplepirate replied:
corrected

thepineapplepirate replied:
thanks for the info! installed vscode extension, should help going forward

thepineapplepirate commented:
Hey @bgruening, 

thanks for the thorough review! I'm going to go over some things and come up with a new draft addressing some issues, particularly what should be a macro and what shouldn't. I will inform you guys when the new draft is pushed back and ready for review. Thanks again for your time and effort with this and the other PRs :).

-BR

bgruening commented:
Cool, thanks!

![grafik](https://user-images.githubusercontent.com/469983/149353685-6ed1e239-6cad-4ab1-8d65-7f43b0aa2041.png)

bgruening commented on tools/buildtools/ambertools/tleap.xml:
```diff
@@ -0,0 +1,1987 @@
+            <param name="arg_zmatrix" label="zmatrix value" type="text" value="" help="LIST"/>
+       </when>
+    </conditional>
+</repeat>
```
indentation ... https://github.com/galaxyproject/galaxy-language-server can help you with that.

thepineapplepirate replied:
corrected

bgruening commented on tools/buildtools/ambertools/tleap.xml:
```diff
@@ -0,0 +1,1987 @@
+        </collection>
+    </outputs>
+    <tests>
+        <test>
```
expected_num_outputs will be helpful here I think

thepineapplepirate replied:
corrected

bgruening commented on tools/buildtools/ambertools/tleap.xml:
```diff
@@ -0,0 +1,1987 @@
+                 <param name="tleap_cmd" value="savePdb" />
+                   <param name="arg_unit" value="mol" />
+            </conditional>
+          </repeat>
```
indentation

thepineapplepirate replied:
corrected

bgruening commented on tools/buildtools/ambertools/tleap.xml:
```diff
@@ -0,0 +1,1987 @@
+                         <has_text text="Welcome to LEaP!" />
+                    </assert_contents>
+               </output>
+               <output_collection name="output_top" type="list">
```
you can also test for the number of elements in this collection, I think with "count"

thepineapplepirate replied:
do you have a link to an example on how to use "count" when testing a collection? thanks man.

simonbray replied:
`<output_collection name="output_top" type="list" count="3">` should work, here are some examples from IUC: https://github.com/galaxyproject/tools-iuc/search?l=XML&q=output_collection+count

thepineapplepirate replied:
Ok, thanks @simonbray!

thepineapplepirate replied:
added count lines

bgruening reviewed (commented):
* would it be useful to also test for the consistency of tleap_in ? I mean that the content is as expected?

thepineapplepirate commented:
> * would it be useful to also test for the consistency of tleap_in ? I mean that the content is as expected?

added a 'lines_diff' to verify consistency in the produced config file
